### PR TITLE
fix(@aws-amplify/auth, amazon-cognito-identity-js): Include clientMetadata for token refresh

### PR DESCRIPTION
**Pull request by elorzafe**
Include client metadata from `Auth.configure` to be send also for token refresh

### Use case that is fixing
```jsx
Auth.configure(awsconfig);
Auth.configure({
  clientMetadata: {
    test: 'this should be on refresh token as well'
  }
})

function App() {
  async function signIn() {
    const user = await Auth.signIn({
      username: 'xxxxx',
      password: 'yyyy',
    })
  }

  async function currentUser() {
    const user = await Auth.currentAuthenticatedUser({ bypassCache: true });
  }
  return (
    <div className="App">
      <header className="App-header">
        <button onClick={signIn}>sign in</button>
        <button onClick={currentUser}>refresh token</button>
      </header>
    </div>
  );
}
```

When refreshing the token when doing `Auth.currentAuthenticatedUser` Amplify is not sending `ClientMetadata` param to Cognito service.

_Issue #, if available:_
#6731 
_Description of changes:_
`CognitoUser` in `amazon-cognito-identity-js` has been added a `clientMetadata` param for `getSession` and `getUserData` functions in order to be used for `refreshSession`.

`Auth` category on `@aws-amplify/auth` has been added `clientMetadata` from `this._config` to be used in `getSession` and `getUserData` operations.

Also added type safety for `CognitoUserPool` type in `Auth` category

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/Auth.ts [hunk 1]
```diff
@@ -662,6 +682,8 @@ export class AuthClass {
 	): Promise<string> {
 		const that = this;
 		return new Promise((res, rej) => {
+			const clientMetadata = this._config.clientMetadata; // TODO: verify behavior if this is override during signIn
```
Comment (amhinson) [inline packages/auth/src/Auth.ts:685]: Did you mean to leave this TODO in here? Is this something you are able to test currently?

**Review by amhinson — APPROVED** [review]
Just one question about the TODOs in here, but other than that, everything looks good to go 👍

File: packages/auth/src/Auth.ts (hunk 1 above)
Comment (elorzafe) [reply packages/auth/src/Auth.ts:685]: The issue there is that we might need to persist `clientMetadata` on the client in order to support `Auth.signIn` parameters

**Review by ericclemmons — APPROVED** [review]
We approved this in bug bash today.

Thanks for teaching me about this feature!

My understanding is that we have 2 major use-cases:

1. `Auth.signIn(...)` for passing metadata for a privileged operations within lambdas.
2. `Amplify.configure(...)` for persisting metadata so that refresh calls behind-the-scenes can provide this data to lambdas.

**Comment by donaldpipowitch** [comment]
FYI: This breaks TypeScript usage in a patch (?) version as the 2nd argument is now mandatory.

image

It's not clear however, if you actually _have_ to do something or if an empty object would be fine as well. If that is the case it would be nice DX, if the 2nd argument would be optional and default to an empty object.
If an empty object is not enough, please make the type more explicit (and maybe do this in a major change?).

Thanks!

**Comment by amhinson** [comment]
@donaldpipowitch Thanks for bringing that to our attention! We've made a fix in #7654 to address this, and it is available on the `unstable` tag now. We will likely do another release to `latest` next week 👍

**Comment by monae** [comment]
@amhinson It seems not just a type declaration failure but a bug introducing breaking change of the method interface.
In our code `user.getSession(callback)` call became to produce runtime error of null reference.
The `options` argument must be optional in js code:

```js
getSession(callback, options = {}) {
  ...
```

**Comment by bluepeter** [comment]
This issue is now resulting in following error for us. (Referencing here as the search for this error was sparse.)

> TypeError : Cannot read property 'clientMetadata' of undefined Please contact us if this error is happening frequently for you.

**Comment by donaldpipowitch** [comment]
Yes. Like @monae  said it's not just a breaking change from a TypeScript perspective, it also results in a runtime error, if you don't update the getSession call.

**Comment by bluepeter** [comment]
This is resulting in all our users repeatedly logged out. Very concerned if such a major bug is still live in PROD. (And assumed to be nothing more than lint?) We've had to revert to an older release. 

@donaldpipowitch does the merged PR for this issue resolve the runtime err?

**Comment by amhinson** [comment]
#7664 Should address the issue here. You can test it now with the `unstable` tag, if you need an immediate workaround. We are planning to get a new release out within the next day or so.

**Comment by elorzafe** [comment]
@bluepeter @donaldpipowitch @monae 

Apologies for the inconvenience. 

We have just release the fix for `amazon-cognito-identity-js` version `4.5.10` that solves this issue.

**Comment by bluepeter** [comment]
Thank you! :)

**Pull request merged**